### ISSUE TITLE
Error handling x middleware regression since 2.1.3

### ISSUE THREAD

**Issue by MattIPv4** [honojs/hono#575]
👋 In versions 2.1.3 and beyond, there is a regression in how errors interact with middleware...

In version 2.1.2 and prior, if an error occurred in a route and `app.onError` was invoked, middleware would still be processed for the error response. This would allow middleware such as the built-in `logger` and built-in `cors` to still process the error response, logging the response or injecting CORS headers.

In version 2.1.3 and since, if an error occurs and `app.onError` responds, middleware does not have a chance to process the response, so middleware such as `logger` cannot log the response, and `cors` cannot inject any CORS headers.

```js
const { Hono } = require('hono');
const { logger } = require('hono/logger');
const { cors } = require('hono/cors');

const app = new Hono();
app.use('*', logger());
app.use('*', cors({ origin: '*' }));

app.get('/error', () => {
  throw new Error('Test error');
});

app.onError((err, ctx) => {
  return ctx.json({
    error: true,
    status: 500,
    message: err.message,
  });
});

const main = async () => {
  const res = await app.request('http://localhost/error');
  console.log([ ...res.headers.entries() ]);
};

main();
```

[RunKit demo: hono@latest](https://runkit.com/mattipv4/6338d5a353c9c20008210cfa)

[RunKit demo: hono@2.1.2](https://runkit.com/mattipv4/6338d71a084d5a0008dd6ed0)

This is pretty breaking for my project, as I would like to have error responses logged, and I'd like them to contain CORS headers like any other response.

**Comment by yusukebe** [comment]
You are right; we have changed the way of error handling. However, this is not a good approach, and I would like to revert back to it, with other changes.

Please wait!

**Comment by yusukebe** [comment]
Hi @MattIPv4 !

In this PR, we make it behaves the same as 2.1.2 and prior. The CORS and logger middleware will work even after the Handler throws an Error.

https://github.com/honojs/hono/pull/576

The new version, including this fix, will be released soon!

**Comment by MattIPv4** [comment]
Awesome, thank you! I look forward to testing it once that release is cut 💙

**Comment by yusukebe** [comment]
Hi! new version has been released:

https://github.com/honojs/hono/releases/tag/v2.2.4

**Comment by MattIPv4** [comment]
Hey! It looks like it's been fixed for synchronous routes :tada:, but the issue still remains for any route that is asynchronous 😢 

[2.1.2](https://runkit.com/mattipv4/633b6d2e74e576000865b3c1)

[2.2.4](https://runkit.com/mattipv4/633b6cf492effa0008bc7450)

**Comment by yusukebe** [comment]
Hi @MattIPv4 !

Sorry! Fixed it and release a new version, "v2.2.5". Please try it and if you still have problems, tell us again!
